docs/proposals/shoot-control-plane-migration/README.md

```diff
@@ -0,0 +1,234 @@
+  * While restoring it should take care of dependencies.
+  * Till the time of restoration, we will freeze backup to ignore updates.
+
+* On container restart, it will take backup of all objects.
+
+##### Gardener modification
+
+* Gardener controller manager component config will have configuration of unique gardener-backup-bucket {can be bucket or private git repo} accessible. This will have backup of backup infrastructure related resources.
+* Gardener will create seed namespace for shoot.
+* It will create backup namespace and restore backup infrastructure related configmaps and secrets. Again restoration will be ignored in case of existing resources with higher revision.
+* Wait until Backup-infrastructure controller create/reconcile backup resources successfully.
+* Gardener will deploy shoot-backup-restore deployment.
+* Gardener call `/initialize` endpoint and will wait until it succeeds.(Logic similar to etcd bootstrap script can be used for optimization.)
+* Gardener will proceed with rest of shoot reconciliation.
+
+#### Extensibility of backup/restore mechanism for the across-region scenario
+
+This approach can be extended to support migration across regions by replicating the object store across regions. Such replication across regions might be required to be setup as a part of the regular backup mechanism to avoid risking the unavailability of the backups in the case of a downtime for the object store in the source seed's region. AWS, Azure and GCP support multi-region object store.
+
+### Migration Co-ordination
+
+Some of the components deployed as part of the shoot control-plane such as those listed below require careful co-ordination during migration to avoid split-brain or other sorts of problems in co-ordination and race conditions.
+
+* Shoot Machine-controller-manager
+
+  The possibility of two machine-control-managers (in source and destination seed clusters) managing the same set of machines should be avoided.
+
+* Shoot Etcd StatefulSet
+
+  The possibility of two etcd statefulsets (in source and destination seed clusters) backing up data to the same object store should be avoided.
+
+* VPN?
+
+#### Landscape Diagram
+
+![Landscape diagram](landscape-diagram.png)
+
+#### Components of the solution
+
+##### Lease Service
+
+The Lease Service will act as the central point of truth for all the seed clusters about which seed owns a particular shoot.
+
+* It needs to be accessible from all the seed clusters among which the shoot control-plane can be moved.
+* It needs to be highly available to reduce chances of false positives/negatives.
+* Given the particular scenario of garden/seed/shoot architecture, shoot YAML in the garden cluster is already the true single point of truth. The Lease Service would be just a secondary source of the truth. Hence, there is no need for any explicit mechanism for lease renewal or leader promotion. That can be done automatically based on the shoot YAML.
+
```

Comment:
Actually, I am a bit scared by a lease service, the HA requirement and its consequences, because if there is a problem, an otherwise healthy control plane could just decide to shut down for good and we could lose all shoots together until the problem gets resolved.

Also, the "factual truth" is elsewhere (not even the shoot resource, which is __only__ the "primary source", but not known to Kubernetes itself and unreachable from the seed/shoot anyhow), it's in the **API server endpoint that the end users and the kubelets use**. So it's in the API server DNS record itself. Everything is determined by it. If it sits wrong, the cluster is dead. Coincidently, DNS is already highly available.

Is there a way to design this thing with less components, founded on battle-proven services (like DNS) and make sure, __something only changes when the Gardener initiates it__, i.e. when it actively decides to move a control plane (rather than autonomously like with the lease service, that, if it has a bug, could shutdown down all shoots even if there is nothing wrong with them - on the other hand, if the DNS record is wrong, the control plane can shutdown - nothing is to be gained from an unreachable control plane anymore).

Reply:
I am also worried about the same things. But I could not readily think of a way to get around the problem of DNS record propagation. Let me see if there is some around that.